Neos.ContentRepository/Classes/Domain/Projection/Content/TraversableNodes.php

```diff
@@ -0,0 +1,147 @@
+<?php
+namespace Neos\ContentRepository\Domain\Projection\Content;
+
+/*
+ * This file is part of the Neos.ContentRepository package.
+ *
+ * (c) Contributors of the Neos Project - www.neos.io
+ *
+ * This package is Open Source Software. For the full copyright and license
+ * information, please view the LICENSE file which was distributed with this
+ * source code.
+ */
+
+use Neos\Flow\Annotations as Flow;
+
+/**
+ * @Flow\Proxy(false)
+ */
+final class TraversableNodes implements \IteratorAggregate, \Countable
+{
+    /**
+     * @var TraversableNodeInterface[]
+     */
+    private $nodes;
+
+    private function __construct(array $nodes)
+    {
+        $this->nodes = $nodes;
+    }
+
+    public static function fromArray(array $nodes): self
+    {
+        foreach ($nodes as $node) {
+            if (!$node instanceof TraversableNodeInterface) {
+                throw new \InvalidArgumentException(sprintf('TraversableNodes only support instances of %s, given: %s', TraversableNodeInterface::class, is_object($node) ? get_class($node): gettype($node)), 1542893076);
+            }
+        }
+        return new static($nodes);
+    }
+
+    public function merge(self $other): self
+    {
+        return new static(array_merge($this->nodes, $other->nodes));
+    }
+
+    public function isEmpty(): bool
+    {
+        return $this->nodes === [];
+    }
+
+    public function count(): int
+    {
+        return count($this->nodes);
+    }
+
+    private function getNodeIndex(TraversableNodeInterface $subject): int
+    {
+        foreach ($this->nodes as $index => $node) {
+            if ($node === $subject) {
+                return $index;
+            }
+        }
+        throw new \InvalidArgumentException('This node does not exist in this set', 1542901216);
```

Comment:
Would be great to have 'this node' in the exceptions somehow specified. What could that be, the NodeAggregateIdentifier?

Reply:
fixed with followup commit :)